disk import: Prioritize returning upload errors

wfchandler opened the pull request:
When sending chunks of the file to upload tasks, we return an error if the channel to that task is closed. This situation will happen if the upload task has panicked or already returned due to an error. However, we check for an error from the sending task first, and the `failed to send chunks` error will be surfaced to the user, rather than the true problem that occurred in the upload task.

In the inverse situation, where the reader fails, the uploaders will return gracefully when their channels are closed, so we do not need to worry about a problem reading the file showing up as an upload error.

Update our error handling to check for an error from the upload tasks, then the reader if those returned cleanly.

wfchandler committed:
disk import: Prioritize returning upload errors

When sending chunks of the file to upload tasks, we return an error if
the channel to that task is closed. This situation will happen if the
upload task has panicked or already returned due to an error. However,
we check for an error from the sending task first, and the `failed to
send chunks` error will be surfaced to the user, rather than the true
problem that occurred in the upload task.

In the inverse situation, where the reader fails, the uploaders will
return gracefully when their channels are closed, so we do not need to
worry about a problem reading the file showing up as an upload error.

Update our error handling to check for an error from the upload tasks,
then the reader if those returned cleanly.

wfchandler commented:
Issue reported by a customer in https://oxidecomputer.slack.com/archives/C089VG8LHPG/p1738198958814399

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -415,18 +415,25 @@ pub mod types {
+            if results.iter().any(|r| r.is_err()) {
+                let mut msg = String::from("upload task(s) failed:");
+                for (i, err) in results.into_iter().filter_map(Result::err).enumerate() {
+                    msg += &format!("\n  task {i}: {err}");
```
I think we want to format these errors differently. First: the numeric order of the errors is unimportant (which makes the fact that we're reporting the nth error as though it were the error from the nth task basically irrelevant to the user). Let's do something a bit more like this:

```rust
let errors = results.into_iter().filter_map(Result::err).collect::<Vec<_>>();
if errors.len() == 1 {
    // You experienced an error and here it is!
} else if errors.len() > 1 {
  // Shit, bruh, more than one thing happened. Here's a list of the things that happened
}
```

wfchandler replied:
Makes sense, I added a leading `*` for the multiple error case, since the error messages are long enough they may wrap in the user's terminal.

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -415,18 +415,25 @@ pub mod types {
 
+            // Check for errors while reading the file. We do second as an error in an
+            // uploading task will cause a failure here, and we should surface the
+            // triggering error to the user.
```
This didn't quite parse for me. I think it's "We do this after checking upload task errors because a failure in an upload task may result in an error where the reading task fails to send a chunk to the appropriate uploading task"

wfchandler replied:
I ended up removing this and moving the read error result back to its original position now that it ignores `send` errors.

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -415,18 +415,25 @@ pub mod types {
                 drop(tx);
```
up at 399 should we be generating an error in that case? If I understand correctly, the call to `send` is only going to fail when the sending task encounters some failure... in which case that task is already going to report a failure. That is to say: should we just ignore send errors?

Here's the docs for send:

> If the receive half of the channel is closed, either due to [close](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.close) being called or the [Receiver](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html) handle dropping, the function returns an error. The error includes the value passed to send.

wfchandler replied:
Yes, good point. We should still `break` on a `send` error since the overall request will fail, but should not treat it as an error for the file reader task.

ahl reviewed (commented):
What if there's a legitimate problem reading a file **and** there's a problem sending a chunk. I'd suggest we might want to look at both the `results` and `read_result` and say: "we had this 1 problem" or "there were several problems and these were they"

wfchandler committed:
Don't list task index in error

The index of the failed task doesn't provide more context to users and
might be accidentally construed as a backtrace. Just list the bare
errors for each failed request.

Also, use a standard `DiskImportError::context` if a single task failed.

wfchandler committed:
Don't treat failure to send as an error

If the file reader task fails to send a chunk to an upload task, we know
this is because that task has either returned due to an error or
panicked. In either case this isn't an error for the reader, which
should return early as the overall request will fail.

wfchandler committed:
Prepend '*' to error

The errors returned by a connection failure are long enough to wrap in
a user's terminal. Add a `*` to the start to make distinguishing errors
easier.

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -417,14 +418,29 @@ pub mod types {
 
             read_result?;
```
I was suggesting we might lump this in with the other errors.

wfchandler replied:
Ah good call, then we'll print all errors to the user if both failed.

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -417,14 +418,29 @@ pub mod types {
+                    }
+                    return Err(DiskImportError::Other(msg.into()));
+                }
+                _ => {}
```
```suggestion
                0 => {}
```

wfchandler replied:
👍

ahl commented on sdk/src/extras/disk.rs:
```diff
@@ -417,14 +418,29 @@ pub mod types {
+            match errors.len() {
+                1 => {
+                    return Err(DiskImportError::context(
+                        "upload task failed",
```
I don't think "task" is meaningful to end users.

I'm imaging something like this:

```
Error while uploading the disk image: Error Response: status: 503 Service Unavailable;
```

OR

```
Errors while uploading the disk image:
* error reading at offset 4096: EIEIO
* block upload failed: 503
* block upload failed: 501c3
```
```

wfchandler replied:
👍

ahl commented on cli/tests/test_disk_import.rs:
```diff
@@ -529,6 +529,79 @@ fn test_disk_import_bad_file_size() {
+        .arg("--disk")
+        .arg("test-disk-import-bulk-import-start-fail")
+        .arg("--parallelism") // Ensure only one write request is sent.
+        .arg("1")
```
can we set this higher to validate how multiple errors are rendered?

wfchandler replied:
Good point, done.

ahl reviewed (commented):
good progress. I think the break Ok(()) is right from the send failure; nice comment.

wfchandler committed:
Include read_result in error summary

wfchandler committed:
Clarify error and test multiple error case
